### PR TITLE
Admin courses

### PR DESCRIPTION
**Pull request by yuchesu**
Finish teaching point calculation logic

### CODE REVIEW

**Comment by jjlock** [comment]
Reviewing this now.

File: management_app/views/points.py [hunk 1]
```diff
@@ -69,12 +67,31 @@ def calculate_teaching_point_val(course_title_id, num_of_enrollment, offload_or_
             y2 = year
             y3 = year
 
-        sql_statement = f"SELECT COUNT(*) FROM scheduled_teaching WHERE ((year = {y1} AND quarter = 1) OR (year = {y2} AND quarter = 2) OR (year = {y3} AND quarter = 3)) AND teaching_point_val = {point_c4}"
-        num_of_row = db.execute(sql_statement).fetchone()         
+        # count by "teaching_point_val" column rather than by "units" column (units = 2) because we only stop giving points when the faculty is already given 0.5 in total
+        sql_statement = f"SELECT teaching_point_val FROM scheduled_teaching st JOIN courses ON st.course_title_id = courses.course_title_id WHERE ((year = {y1} AND quarter = 1) OR (year = {y2} AND quarter = 2) OR (year = {y3} AND quarter = 3)) AND units = 2 AND user_id = {user_id}"
+        rows = db.execute(sql_statement).fetchall()
```
Comment (jjlock) [inline management_app/views/points.py:72]: Using format strings to create a SQL statement is vulnerable to SQL injection. Use `db.execute` to create the SQL statement.

**Review by jjlock — CHANGES_REQUESTED** [review]

File: management_app/views/points.py [hunk 2]
```diff
@@ -38,7 +38,7 @@ def calculate_teaching_point_val(course_title_id, num_of_enrollment, offload_or_
     ).fetchone()
```
Comment (yrfang) [inline management_app/views/points.py]: I was wondering if we need to check `row` is None when there is no matched `course_title_id` when uploading the teaching_schedule file? Only when there is a valid course_title_id to match teaching_schedule, we process 'units' and 'course_level' to update enrollment and further calculate the new `teaching_point_val`?

File: management_app/views/course.py [hunk 3]
```diff
@@ -101,74 +120,249 @@ def upload_user_file():
...
-        #   I need to insert all data in the scheduled_teaching beforhand  
-        #   (rule: Points are divided equally between the instructors for a co-taught course.)
+            insert_or_update_scheduled_teaching_for_each_user(user_UCINetID_list, num_of_enrollment, user_id_and_academic_year_set, academic_year, combine_with, rows_dict, course_title_id, year, quarter, course_sec, offload_or_recall_flag, num_of_co_taught)            
 
-        # TODO: call calculate professor's point API (written by ying-ru): update_yearly_ending_balance(user_id, year, is_recursive) ??   (send academic year to ying-ru)
-                    
+        flash('Upload scheduled teaching file successfully!', 'success')
+
+        # rules #5: Combined grad/undergraduate classes
+        calculate_combined_classes_and_update_scheduled_teaching(rows_dict)
```
Comment (yrfang) [inline management_app/views/course.py:163]: Somehow when I uploaded a new schedule_teaching_point file with a new enrollment number, I had an issue with the `TypeError: 'dict_keys' object is not subscriptable` error. Let's check this a bit offline.

**Review by yrfang — CHANGES_REQUESTED** [review]

File: management_app/views/course.py (hunk 3 above)
Comment (yrfang) [reply management_app/views/course.py:163]: For `calculate_combined_classes_and_update_scheduled_teaching` function, would it call when`rows_dict` is not empty? So it is feasible to process keys from `rows_dict.keys()`.

**Review by yrfang — APPROVED** [review]
The key error of `rows_dict` has been fixed. Verified without and with uploaded enrollment data from the local branch, but for the correct ending balance point, will verify in testing tasks.
```
keys = list(rows_dict.keys())
```

**Comment by yrfang** [comment]
For SQL injection, the new commit includes validation to prevent injection vulnerability, so I'm going to merge this PR.

**Pull request merged**